### PR TITLE
Handle only known functions in rm_rf

### PR DESCRIPTION
**Pull request by blueyed**
Ref: https://github.com/pytest-dev/pytest/issues/5626

### CODE REVIEW

File: src/_pytest/pathlib.py [hunk 1]
```diff
@@ -64,8 +70,7 @@ def force_writable_and_retry(function, p, excinfo):
 
         chmod_w(p)
         try:
-            # retry the function that failed
-            function(str(p))
+            shutil.rmtree(str(p))
```
Comment (blueyed) [inline src/_pytest/pathlib.py]: This needs to be special cased for dir/file (TestNumberedDir.test_rmtree_with_read_only_directory fails).

**Comment by blueyed** [comment]
@nicoddemus 
Feel free to pick this up.  I am on `features` for now.. ;)

**Comment by nicoddemus** [comment]
> Feel free to pick this up. I am on features for now.. ;)

What do you mean, has this been fixed in `features`? I missed that

**Comment by blueyed** [comment]
No - it's not in features yet.

**Comment by nicoddemus** [comment]
Sorry I'm a little confused, when you said "I'm on features now", I understood that you were using pytest@features and this was no longer an issue for you. Did you mean something else?

**Comment by blueyed** [comment]
No, I've meant that: https://github.com/pytest-dev/pytest/issues/5626 is only an issue on master, not features (since you've merged it to master).
IIRC it is not released yet (and master was not merged into features - I was thinking about this initially, but figured it would be better to fix it on master).

**Comment by nicoddemus** [comment]
Ahh got it, thanks. 👍 

I'll pick this up, probably will copy `pre-commit`'s solution given that this is probably already battle-tested.

**Comment by nicoddemus** [comment]
(Taking in account that your issue was really with `os.open`

**Comment by nicoddemus** [comment]
@blueyed moved the internal error function to the module level and added a unittest. I think this is the best we can do, as testing the actual cases would involve a lot of work.

Please rebase/merge if you think this is good. 👍

**Comment by blueyed** [comment]
@nicoddemus 
Looks decent, thanks!

**Comment by RonnyPfannschmidt** [comment]
looks awesome, please sort the type error on py35

**Comment by nicoddemus** [comment]
Should be ready, @blueyed please squash/merge when you have a chance, thanks!

**Comment by nicoddemus** [comment]
After this, I plan to port this to `4.6-maintenance`, and prepare `5.1` and `4.6.5`. 🎉

**Comment by nicoddemus** [comment]
@blueyed let me know if you want me to squash this myself.

**Comment by blueyed** [comment]
@nicoddemus 
Please go ahead - I won't have time for this today.

**Comment by nicoddemus** [comment]
Done. This should be merged as soon as CI passes.

**Pull request merged**